src/ServiceControl.Transports.AzureStorageQueues/QueueLengthProvider.cs

```diff
@@ -0,0 +1,134 @@
+﻿namespace ServiceControl.Transports.SQLServer
+{
+    using System;
+    using System.Collections.Concurrent;
+    using System.Linq;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using Microsoft.WindowsAzure.Storage;
+    using Microsoft.WindowsAzure.Storage.Queue;
+    using Monitoring;
+    using Monitoring.Infrastructure;
+    using Monitoring.Messaging;
+    using Monitoring.QueueLength;
+    using NServiceBus.Logging;
+    using NServiceBus.Metrics;
+
+    public class QueueLengthProvider : IProvideQueueLength
+    {
+        ConcurrentDictionary<EndpointInputQueue, CloudQueue> queues = new ConcurrentDictionary<EndpointInputQueue, CloudQueue>();
+        ConcurrentDictionary<CloudQueue, int> sizes = new ConcurrentDictionary<CloudQueue, int>();
+        ConcurrentDictionary<CloudQueue, CloudQueue> problematicQueues = new ConcurrentDictionary<CloudQueue, CloudQueue>();
+
+        string connectionString;
+        QueueLengthStore store;
+
+        CancellationTokenSource stop = new CancellationTokenSource();
+        Task pooler;
+
+        public void Initialize(string connectionString, QueueLengthStore store)
+        {
+            this.connectionString = connectionString;
+            this.store = store;
+        }
+
+        public void Process(EndpointInstanceId endpointInstanceId, EndpointMetadataReport metadataReport)
+        {
+            var endpointInputQueue = new EndpointInputQueue(endpointInstanceId.EndpointName, metadataReport.LocalAddress);
+            var queueName = metadataReport.LocalAddress.ToLower().Replace(".", "-");
```

Comment:
@Particular/azure-maintainers Could you take a look? It's been a while since I worked with ASQ and I don't like this naming transformation, but it works. If you have anything that I could reuse/apply to make it better, I'd appreciate any proposal.

Reply:
That doesn't look right.
Is `metadataReport.LocalAddress` value has gone thought the transports' `ToTransportAddress` transformation?

Also as of version [7.4.2](https://github.com/Particular/NServiceBus.AzureStorageQueues/blob/7.4.2/src/Transport/AzureStorageQueueInfrastructure.cs#L181) `ToTransportAddress()` was fixed to convert `LogicalAddress` with all the rules and registered [shortening strategy](https://github.com/Particular/NServiceBus.AzureStorageQueues/blob/7.4.2/src/Transport/QueueAddressGenerator.cs). Prior to that, (7.0.0 - [7.4.1](https://github.com/Particular/NServiceBus.AzureStorageQueues/blob/7.4.1/src/Transport/AzureStorageQueueInfrastructure.cs#L187)) the transport had shortening logic spread in various places such as [dispatcher](https://github.com/Particular/NServiceBus.AzureStorageQueues/blob/7.4.1/src/Transport/Dispatcher.cs#L60),  🙁

Reply:
> OK, so I checked and we are not using QueueBindings. We are instead using .LocalAddress(). That address has been through the transport address sanitization process so it will be the actual physical address.

If this is true, then there's going to be a discrepency between versions 7.4.2 and higher and 7.4.1 and lower based on what I've posted above. Will raise this in the TF channel for discussion as well.

Reply:
@SeanFeldman looking at the address [generator](https://github.com/Particular/NServiceBus.AzureStorageQueues/blob/7.4.2/src/Transport/QueueAddressGenerator.cs) the `GetQueueName` is idempotent. So it should be fine to run this logic at the SC side one more time no matter which version of the endpoint generated the `LocalAddress`.